### PR TITLE
WIP: Aws schema

### PR DESCRIPTION
**Pull request by kornypoet**
Since we seem to be actively discussing this, I present the AWS schema change as a _work in progress_. 

Important things to note:
* I moved a couple of 'options' to `cfndsl/global.rb` and adjusted the load order. This allows both the executable _and_ ruby scripts to affect cfndsl's behavior before types are created and templates are evaluated.
* As outlined in stevenjack/cfndsl#276 the behavior is as follows:
  * any invocation of cfndsl will download the latest schema to a specified location (with a sane default) if it does not exist.
  * Updating this file is allowable, and exits cleanly
  * Using a different schema file when converting templates is possible
* Lastly, the missing piece of code (still to be written) will go in `cfndsl/types.rb`, and it will ingest the schema file (AWS only) and parse through it and recreate it exactly as the current code specifies. No altered behavior in this iteration. As we move forward, we can add things like validations.

### CODE REVIEW

**Comment by gergnz** [comment]
New thought.
Is us-east-1 a superset?  Are there any services not available in us-east-1? If so it would be more advantageous to pull them all and merge. But super ugly!

**Comment by kornypoet** [comment]
According to this, no: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html

**Comment by allinwonder** [comment]
I think we can pick one region to start and have a conversation with AWS to request a superset endpoint is possible

**Comment by cmaxwellau** [comment]
All services are generally released in us-east-1 first, so using the schema
from this region ought to include everything.

**Comment by kornypoet** [comment]
Interesting wrinkle: the AWS published schemas have errors in them; things like undefined types, incorrect capitalization, and inconsistent pluralization.

**Comment by elmobp** [comment]
ouchie

**Comment by gergnz** [comment]
I spoke to @domdivakaruni and he's keen to get some details and/or have a chat about the issues @kornypoet

**Comment by kornypoet** [comment]
Absolutely. Through what channels? Github issues, slack, email?

**Comment by kornypoet** [comment]
@gergnz So, I've finally been able to spend some time trying to wrangle this together. This is a large change and I'd really like your feedback. A couple points:
* This is a transition. The important part is being able to define types using the resource specification. All original plumbing is left in their pending the big 1.0.0 release
*  I've tried to comment all of the logic as much as possible. Feel free to ask any questions about things you don't understand.
* There is a cli test failing b/c of the new download logic, I need to look into that
* This new version works seamlessly with all of my company's existing templates. If you find some of yours that don't paste me broken snippets.
* Lastly, thank you in advance 😄

File: lib/cfndsl/specification.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+          end
+          extracted[nested_prop_name] = nested_prop_type
+          extracted
+        end
+        types[property_name] = properties
+        types
+      end
+    end
+
+    # Missing/malformed resources from the resource specification
```
Comment (elmobp) [inline lib/cfndsl/specification.rb]: @gergnz @kornypoet could I suggest moving this to a more common area like we did for fixing Parameter and Output?

**Comment by elmobp** [comment]
SSM::Parameter not in the spec either see below:
```bash
tmp michaelbaker$ cat Gemfile
source 'https://rubygems.org'
gem 'cfndsl', :git => 'https://github.com/kornypoet/cfndsl.git', :branch => 'aws-schema'
tmp michaelbaker$ cat test.rb 
require 'cfndsl'

CloudFormation do
  Description 'test'

  SSM_Parameter('test') do
   Name 'test'
   Type 'String'
   Value 'date'
   Description 'test'
  end
end
tmp michaelbaker$ bundle exec cfndsl test.rb
The creation of constants as config is deprecated!
Please switch to the #external_parameters method within your templates to access variables
See https://github.com/stevenjack/cfndsl/issues/170
Use the --disable-binding flag to suppress this message
bundler: failed to load command: cfndsl (/Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bin/cfndsl)
NoMethodError: undefined method `SSM_Parameter' for #<CfnDsl::CloudFormationTemplate:0x007fe4cb0a01b0>
  /Users/michaelbaker/Documents/Clients/BulletProof/cloudformation/UrbisPro/test.rb:6:in `block in eval_file_with_extras'
  /Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/cfndsl-28adf007259d/lib/cfndsl/jsonable.rb:199:in `instance_eval'
  /Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/cfndsl-28adf007259d/lib/cfndsl/jsonable.rb:199:in `declare'
  /Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/cfndsl-28adf007259d/lib/cfndsl.rb:90:in `CloudFormation'
  /Users/michaelbaker/Documents/Clients/BulletProof/cloudformation/UrbisPro/test.rb:3:in `eval_file_with_extras'
  /Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/cfndsl-28adf007259d/lib/cfndsl.rb:84:in `eval'
  /Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/cfndsl-28adf007259d/lib/cfndsl.rb:84:in `eval_file_with_extras'
  /Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bundler/gems/cfndsl-28adf007259d/bin/cfndsl:107:in `<top (required)>'
  /Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bin/cfndsl:23:in `load'
  /Users/michaelbaker/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bin/cfndsl:23:in `<top (required)>'
tmp michaelbaker$ 
```

**Comment by gergnz** [comment]
so I wrote a quick comparison:
```
"In DSL types"
[
    [0] "AWS::EC2::VPNGatewayConnection",
    [1] "AWS::SSM::Parameter",
    [2] "AWS::Serverless::API",
    [3] "AWS::Serverless::Function"
]
"In Cloudformation"
[
    [ 0] "AWS::ApiGateway::UsagePlan",
    [ 1] "AWS::CodeBuild::Project",
    [ 2] "AWS::CodeCommit::Repository",
    [ 3] "AWS::CodePipeline::CustomActionType",
    [ 4] "AWS::CodePipeline::Pipeline",
    [ 5] "AWS::Cognito::IdentityPool",
    [ 6] "AWS::Cognito::IdentityPoolRoleAttachment",
    [ 7] "AWS::Cognito::UserPool",
    [ 8] "AWS::Cognito::UserPoolClient",
    [ 9] "AWS::Cognito::UserPoolGroup",
    [10] "AWS::Cognito::UserPoolUser",
    [11] "AWS::Cognito::UserPoolUserToGroupAttachment",
    [12] "AWS::DataPipeline::Pipeline",
    [13] "AWS::EC2::Host",
    [14] "AWS::EC2::NetworkInterfaceAttachment",
    [15] "AWS::EC2::PlacementGroup",
    [16] "AWS::EC2::SubnetCidrBlock",
    [17] "AWS::EC2::VPCCidrBlock",
    [18] "AWS::EMR::Cluster",
    [19] "AWS::EMR::InstanceGroupConfig",
    [20] "AWS::EMR::Step",
    [21] "AWS::ElasticBeanstalk::ApplicationVersion",
    [22] "AWS::ElasticBeanstalk::ConfigurationTemplate",
    [23] "AWS::GameLift::Alias",
    [24] "AWS::GameLift::Build",
    [25] "AWS::GameLift::Fleet",
    [26] "AWS::IoT::Certificate",
    [27] "AWS::IoT::Policy",
    [28] "AWS::IoT::PolicyPrincipalAttachment",
    [29] "AWS::IoT::Thing",
    [30] "AWS::IoT::ThingPrincipalAttachment",
    [31] "AWS::IoT::TopicRule",
    [32] "AWS::KinesisFirehose::DeliveryStream",
    [33] "AWS::Lambda::Alias",
    [34] "AWS::Lambda::Version",
    [35] "AWS::Logs::Destination",
    [36] "AWS::Logs::LogStream",
    [37] "AWS::Logs::SubscriptionFilter",
    [38] "AWS::OpsWorks::UserProfile",
    [39] "AWS::OpsWorks::Volume",
    [40] "AWS::SNS::Subscription",
    [41] "AWS::SSM::Association",
    [42] "AWS::WorkSpaces::Workspace"
]
```

**Comment by kornypoet** [comment]
@elmobp Looking into why that one is missing.
@gergnz the `types.yaml` file and the new types defined from the `resource_specification.json` are definitely going to have differences. What I'm hoping to see is that the resource_spec side is the superset of what we had in the `types.yaml` as I assume Amazon updates more frequently than we have in the past. Your comparison: what is it showing that is missing and from what side?

**Comment by kornypoet** [comment]
@elmobp So if you search the `resource_specification.json` file, you'll see that `AWS::SSM::Parameter` isn't even defined ☹️  (and that's from Amazon). I'll add it to the patch. Also, I can move the patches to their own file.

**Comment by kornypoet** [comment]
@gergnz the patches have been updated, and the 4 resources you listed should be present now.
@elmobp patches have been moved to their own file
Also, the cli test now passes 😄

**Comment by gergnz** [comment]
Looking good. I've clarified my description. If I get time I'll write a more complete comparison (properties, and property types, and I'll run this branch through some more templates.

Thanks for the hard work @kornypoet

**Comment by gergnz** [comment]
I add more functionality to my comparison, and these are the ones we need to look out for (as you've stated before @kornypoet )
```
"This resource has differences: AWS::ElasticLoadBalancingV2::Listener"
[
    [0] "SslPolicy"
]
[
    [0] "SSLPolicy"
```

**Comment by kornypoet** [comment]
According to this [documention](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-listener.html), the property name is supposed to be `SslPolicy` (lowercase). That matches the `resource_specification.json` and what is dynamically generated currently.

**Comment by gergnz** [comment]
For reference:
* EIP [AWS::EC2::EIPAssociation](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-eip-association.html#cfn-ec2-eipassociation-eip)
* RoleARN [AWS::Config::ConfigurationRecorder](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-config-configurationrecorder.html#cfn-config-configurationrecorder-rolearn)
* RamdiskId [AWS::EC2::Instance](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html#cfn-ec2-instance-ramdiskid)

So only EIP and RoleARN are incorrect in the official specification.

**Comment by elmobp** [comment]
@kornypoet myself and @gergnz have reached out to AWS about alot of these issues with the schema

**Comment by kornypoet** [comment]
Awesome. At a minimum, if they could point us at a JIRA, or github repository to make PRs against, I'd be super pleased.

**Comment by elmobp** [comment]
@kornypoet we asked for git / crowd sourced

File: lib/cfndsl/patches.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
+module CfnDsl
```
Comment (elmobp) [inline lib/cfndsl/patches.rb]: This now makes sense! good effort!!!

**Comment by kornypoet** [comment]
@elmobp @gergnz Any word from AWS? Also, how are we feeling about this PR?

**Comment by elmobp** [comment]
Sorry @kornypoet I forgot to reply! AWS said there was a new schema being published Friday (I have confirmed this!) that included SSM::Parameter. We have also submitted the feature request to AWS will let you know when I hear more

```bash
ALT-C02Q73JUG8W:tmp michaelbaker$ wc -l CloudFormationResourceSpecification.json 
   15781 CloudFormationResourceSpecification.json
ALT-C02Q73JUG8W:tmp michaelbaker$ wc -l updated.json 
   16516 updated.json
ALT-C02Q73JUG8W:tmp michaelbaker$ 
```

I dont know of a changelog for it though

**Comment by elmobp** [comment]
As for the PR it looks good to me! maybe update docs a lil to say something along the lines of "If your resource cannot be found try updating". I might even given it a stab of throwing some rescue handlers to make the errors look a bit nicer that people can understand easier?

**Comment by gergnz** [comment]
V 1.4.0 dropped the other day from AWS.

We still need to keep our 5 custom resource types, as my comparison tool, still shows them as different. No further word at this stage.

**Comment by gergnz** [comment]
Can we please dump this too?
https://github.com/stevenjack/cfndsl/blob/master/cfndsl.gemspec#L24 ?

**Comment by gergnz** [comment]
Ran my test suite over, and had one oddity:

```patch
-      NotificationConfiguration(
+      NotificationConfigurations(
```

I have no idea how that worked in the past, but it did. Otherwise everything else passed. 👯

**Comment by gergnz** [comment]
Can we merge this against 1.0.0-pre instead of master maybe?

**Comment by kornypoet** [comment]
I was initially thinking of this being the last merge we make to old master, before all work shifts to `1.0.0.`. For that reason, I used the resource_specification to create the dsl methods in the exact same way as before, basically using it as a drop-in replacement for the `aws/types.yaml`. For `1.0.0` I'd like to go further, and combine these large sections of metaprogramming and parsing and clean them up. At that time we could open up the possibility of adding argument validation, required parameters and more.

Given that, if you would still like this merged against 1.0.0, I think I will port it over in pieces. I've been preparing a lot of fast and furious PRs for 1.0.0, and I could incorporate this.

**Comment by gergnz** [comment]
OK, That makes sense. I'm happy for you to merge this then and release 0.13!

**Pull request merged**